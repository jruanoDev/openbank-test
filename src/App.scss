@import "styles/base";

.App {
  background-color: #f3f3f3;
  position: relative;
  min-height: 100%;
  &-header {
    position: absolute;
    background-color: #e3e9e9;
    width: 100%;
    height: 70px;
    z-index: 10;
  }
  &-content {
    padding: 20px 10px;
  }
}
