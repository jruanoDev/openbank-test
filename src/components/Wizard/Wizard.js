import "./Wizard.scss";

import React from "react";

import Stepper from "../../components/Stepper/Stepper";
import useWizardController from "../../hooks/useWizardController/useWizardController";

const Wizard = ({ children }) => {
  const totalSteps = children.length;

  const [index, goForward, goBackwards, restart] = useWizardController({
    index: 0,
    totalSteps,
  });

  const props = { goForward, goBackwards, restart };

  return (
    <div className="Wizard">
      <div className="Wizard-header">
        <Stepper totalSteps={totalSteps} activeStep={index} />
      </div>
      <div className="Wizard-content">
        {React.cloneElement(children[index], props)}
      </div>
    </div>
  );
};

export default Wizard;
