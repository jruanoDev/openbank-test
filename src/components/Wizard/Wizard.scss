@import "../../styles/base.scss";

.Wizard {
  width: 100%;
  max-width: 950px;
  background-color: $white;
  position: relative;
  z-index: 20;
  margin: 0 auto;
  &-header {
    background-color: #ccd5da;
    height: 80px;
    display: flex;
    align-items: center;
    justify-content: center;
  }
  &-content {
    box-shadow: 0 0 3px 1px rgba(0, 0, 0, 0.2);

    & .ProductInfo,
    .Feedback,
    .Form {
      &-content {
        padding: 35px 50px;
        @include sm {
          padding: 35px 15px;
        }
      }
    }
  }
}
