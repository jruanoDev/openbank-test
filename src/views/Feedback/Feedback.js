import "./Feedback.scss";

import React from "react";
import { Col, Container, Row } from "react-grid-system";
import { useTranslation } from "react-i18next";

import { ReactComponent as ArrowRight } from "../../assets/img/arrow_right.svg";
import { ReactComponent as DoneIcon } from "../../assets/img/done.svg";
import { ReactComponent as WarningIcon } from "../../assets/img/warning.svg";
import Button from "../../components/Button/Button";
import ButtonContainer from "../../components/ButtonContainer/ButtonContainer";

const Feedback = ({ success, restart }) => {
  const { t } = useTranslation();

  const okData = {
    title: t("Feedback_Ok_Title"),
    text: t("Feedback_Ok_Text"),
  };

  const koData = {
    title: t("Feedback_Ko_Title"),
    text: t("Feedback_Ko_Text"),
  };

  const onFormCompleted = () => alert(t("Form_Alert"));

  const endAddornment = <ArrowRight fill="#ff0049" width="24" height="24" />;
  const button = success ? (
    <Button
      endAddornment={endAddornment}
      color="primary"
      onClick={onFormCompleted}
    >
      {t("Feedback_Ok_Button")}
    </Button>
  ) : (
    <Button endAddornment={endAddornment} color="primary" onClick={restart}>
      {t("Feedback_Ko_Button")}
    </Button>
  );

  const data = success ? okData : koData;
  return (
    <div className="Feedback">
      <div className="Feedback-content">
        <Container fluid>
          <Row>
            <Col xs="content">
              {success ? (
                <DoneIcon width="64" height="64" />
              ) : (
                <WarningIcon width="64" height="64" />
              )}
            </Col>
            <Col>
              <Row gutterWidth={0}>
                <Col>
                  <h1>{data.title}</h1>
                </Col>
              </Row>
              <Row>
                <Col>
                  <p>{data.text}</p>
                </Col>
              </Row>
            </Col>
          </Row>
        </Container>
      </div>
      <ButtonContainer submitButton={button} />
    </div>
  );
};

export default Feedback;
