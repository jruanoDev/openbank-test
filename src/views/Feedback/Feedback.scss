@import "../../styles/base.scss";

.Feedback {
  &-content {
    padding: 35px 50px;
    & h1 {
      margin: 0;
    }
    & p {
      margin-top: 10px;
    }
  }
}
