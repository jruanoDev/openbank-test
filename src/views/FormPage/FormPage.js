import PropTypes from "prop-types";
import React, { useState } from "react";
import { Col, Container, Row } from "react-grid-system";
import { Trans, useTranslation } from "react-i18next";

import { ReactComponent as ArrowRight } from "../../assets/img/arrow_right.svg";
import Button from "../../components/Button/Button";
import ButtonContainer from "../../components/ButtonContainer/ButtonContainer";
import { submitForm } from "../../services/api";
import PasswordForm from "./PasswordForm/PasswordForm";

const FormPage = ({ onFormSuccess, goForward, goBackwards }) => {
  const [formValues, setFormValues] = useState({
    password: "",
    repeatPassword: "",
    hint: "",
  });

  const { t } = useTranslation();

  const [errors, setErrors] = useState({});
  const [loading, setLoading] = useState(false);

  const errorMessages = {
    password: t("FORM_Password_Validation"),
    repeatPassword: t("FORM_RepeatPassword_Validation"),
  };

  const validate = {
    password: (password) =>
      Boolean(password.match(/^(?=.*[0-9])(?=.*[A-Z])(?=.{8,24}$)/))
        ? undefined
        : errorMessages.password,
    repeatPassword: (password, repeatPassword) =>
      password === repeatPassword ? undefined : errorMessages.repeatPassword,
  };

  const handleFormValuesChange = (e) => {
    const { name, value } = e.target;

    setFormValues({
      ...formValues,
      [name]: value,
    });
  };

  const handleFormSubmit = () => {
    const { password, repeatPassword } = formValues;

    const formValidation = {
      password: validate.password(password),
      repeatPassword: validate.repeatPassword(password, repeatPassword),
    };

    const end = (success) => {
      onFormSuccess(success);
      goForward();
    };

    if (!formValidation.password && !formValidation.repeatPassword) {
      setErrors({});
      setLoading(true);

      submitForm(password)
        .then(() => {
          setLoading(false);
          end(true);
        })
        .catch(() => {
          setLoading(false);
          end(false);
        });
    } else {
      setErrors(formValidation);
    }
  };

  return (
    <div className="Form">
      <div className="Form-content">
        <Container fluid>
          <Row>
            <Col>
              <h1 className="h1-sub">{t("PasswordManagerTitle")}</h1>
            </Col>
          </Row>
          <Row>
            <Col>
              <p>
                <Trans i18nKey="FORM_Description">
                  En primer lugar, debes crear una contraseña diferente para sus
                  pertenencias electrónicas. <br />
                  No podrás recuperar tu contraseña, así que recuérdela bien.
                </Trans>
              </p>
            </Col>
          </Row>
          <PasswordForm
            values={formValues}
            onChange={handleFormValuesChange}
            errors={errors}
          />
        </Container>
      </div>
      <ButtonContainer
        submitButton={
          <Button
            type="filled"
            endAddornment={<ArrowRight fill="#fff" width="24" height="24" />}
            onClick={handleFormSubmit}
            loading={loading}
          >
            {t("Wizard_Next_Button")}
          </Button>
        }
        cancelButton={
          <Button onClick={goBackwards}>{t("Wizard_Back_Button")}</Button>
        }
      />
    </div>
  );
};

FormPage.propTypes = {
  index: PropTypes.number,
  goForward: PropTypes.func,
  goBackwards: PropTypes.func,
  restart: PropTypes.func,
  onFormSuccess: PropTypes.func.isRequired,
};

export default FormPage;
