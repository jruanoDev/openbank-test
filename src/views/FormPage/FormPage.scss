@import "../../styles/base.scss";

.Form {
  &-content {
    padding: 35px 50px;
  }
}
