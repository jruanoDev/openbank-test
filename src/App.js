import "./App.scss";

import React, { useState } from "react";

import Wizard from "./components/Wizard/Wizard";
import Feedback from "./views/Feedback/Feedback";
import FormPage from "./views/FormPage/FormPage";
import ProductInformation from "./views/ProductInformation/ProductInformation";

const App = () => {
  const [success, setSuccess] = useState(false);

  return (
    <main className="App">
      <header className="App-header" />
      <section className="App-content">
        <Wizard>
          <ProductInformation />
          <FormPage onFormSuccess={(o) => setSuccess(o)} />
          <Feedback success={success} />
        </Wizard>
      </section>
    </main>
  );
};

export default App;
