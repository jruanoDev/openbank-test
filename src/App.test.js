import { shallow } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";

import App from "./App";
import Wizard from "./components/Wizard/Wizard";
import Feedback from "./views/Feedback/Feedback";
import FormPage from "./views/FormPage/FormPage";
import ProductInformation from "./views/ProductInformation/ProductInformation";

describe("App", () => {
  it("renders without crashing", () => {
    shallow(<App />);
  });

  it("snapshot renders", () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders Wizard component", () => {
    const component = shallow(<App />);
    expect(component.find(Wizard).length).toEqual(1);
  });

  it("renders Wizard component with 3 steps", () => {
    const component = shallow(<App />);
    const wizard = component.find(Wizard);
    expect(wizard.find(ProductInformation).length).toEqual(1);
    expect(wizard.find(FormPage).length).toEqual(1);
    expect(wizard.find(Feedback).length).toEqual(1);
  });
});
