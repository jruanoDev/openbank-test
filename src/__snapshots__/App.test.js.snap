// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
<main
  className="App"
>
  <header
    className="App-header"
  />
  <section
    className="App-content"
  >
    <div
      className="Wizard"
    >
      <div
        className="Wizard-header"
      >
        <div
          className="Stepper"
        >
          <div
            className="Step Step-active"
          >
            1
          </div>
          <div
            className="Step-bar Step-bar-active"
          />
          <div
            className="Step Step"
          >
            2
          </div>
          <div
            className="Step-bar Step-bar"
          />
          <div
            className="Step Step"
          >
            3
          </div>
        </div>
      </div>
      <div
        className="Wizard-content"
      >
        <div
          className="ProductInfo"
        >
          <div
            className="ProductInfo-content"
          >
            <div
              style={
                Object {
                  "boxSizing": "border-box",
                  "marginLeft": "auto",
                  "marginRight": "auto",
                  "paddingLeft": 15,
                  "paddingRight": 15,
                  "position": "relative",
                }
              }
            >
              <div
                style={
                  Object {
                    "alignItems": "normal",
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "flexWrap": "wrap",
                    "justifyContent": "flex-start",
                    "marginLeft": -15,
                    "marginRight": -15,
                  }
                }
              >
                <div
                  style={
                    Object {
                      "boxSizing": "border-box",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "left": "auto",
                      "marginLeft": "0%",
                      "maxWidth": "100%",
                      "minHeight": 1,
                      "paddingLeft": 15,
                      "paddingRight": 15,
                      "position": "relative",
                      "right": "auto",
                      "width": "100%",
                    }
                  }
                >
                  <h1
                    className="h1-sub"
                  >
                    Create your Password Manager
                  </h1>
                </div>
              </div>
              <div
                className="Features"
                style={
                  Object {
                    "alignItems": "normal",
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "flexWrap": "wrap",
                    "justifyContent": "flex-start",
                    "marginLeft": -15,
                    "marginRight": -15,
                  }
                }
              >
                <div
                  className="Features-column"
                  style={
                    Object {
                      "boxSizing": "border-box",
                      "flexBasis": "50%",
                      "flexGrow": 0,
                      "flexShrink": 0,
                      "left": "auto",
                      "marginLeft": "0%",
                      "maxWidth": "50%",
                      "minHeight": 1,
                      "paddingLeft": 15,
                      "paddingRight": 15,
                      "position": "relative",
                      "right": "auto",
                      "width": "50%",
                    }
                  }
                >
                  <svg
                    height={150}
                  >
                    icon2.svg
                  </svg>
                  <p>
                    Save here all your passwords, data or any information, forget paper notes and unprotected applications
                  </p>
                </div>
                <div
                  className="Features-column"
                  style={
                    Object {
                      "boxSizing": "border-box",
                      "flexBasis": "50%",
                      "flexGrow": 0,
                      "flexShrink": 0,
                      "left": "auto",
                      "marginLeft": "0%",
                      "maxWidth": "50%",
                      "minHeight": 1,
                      "paddingLeft": 15,
                      "paddingRight": 15,
                      "position": "relative",
                      "right": "auto",
                      "width": "50%",
                    }
                  }
                >
                  <svg
                    height={150}
                  >
                    icon1.svg
                  </svg>
                  <p>
                    Create your master password, only you will have access to your secrets with it
                  </p>
                </div>
              </div>
              <div
                style={
                  Object {
                    "alignItems": "normal",
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "flexWrap": "wrap",
                    "justifyContent": "flex-start",
                    "marginBottom": 20,
                    "marginLeft": -15,
                    "marginRight": -15,
                  }
                }
              >
                <div
                  style={
                    Object {
                      "boxSizing": "border-box",
                      "flexBasis": "100%",
                      "flexGrow": 0,
                      "flexShrink": 0,
                      "left": "auto",
                      "marginLeft": "0%",
                      "maxWidth": "100%",
                      "minHeight": 1,
                      "paddingLeft": 15,
                      "paddingRight": 15,
                      "position": "relative",
                      "right": "auto",
                      "width": "100%",
                    }
                  }
                >
                  <h3>
                    How it works?
                  </h3>
                </div>
                <div
                  style={
                    Object {
                      "boxSizing": "border-box",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "left": "auto",
                      "marginLeft": "0%",
                      "maxWidth": "100%",
                      "minHeight": 1,
                      "paddingLeft": 15,
                      "paddingRight": 15,
                      "position": "relative",
                      "right": "auto",
                      "width": "100%",
                    }
                  }
                >
                  <p
                    style={
                      Object {
                        "marginTop": 10,
                      }
                    }
                  >
                    First, you must create a different password for your electronic belongings. You won't be able to recover your password, so remember it well.
                  </p>
                </div>
              </div>
              <div
                style={
                  Object {
                    "alignItems": "normal",
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "flexWrap": "wrap",
                    "justifyContent": "flex-start",
                    "marginLeft": -15,
                    "marginRight": -15,
                  }
                }
              >
                <div
                  style={
                    Object {
                      "boxSizing": "border-box",
                      "flexBasis": "100%",
                      "flexGrow": 0,
                      "flexShrink": 0,
                      "left": "auto",
                      "marginLeft": "0%",
                      "maxWidth": "100%",
                      "minHeight": 1,
                      "paddingLeft": 15,
                      "paddingRight": 15,
                      "position": "relative",
                      "right": "auto",
                      "width": "100%",
                    }
                  }
                >
                  <h3>
                    What type of data can you save?
                  </h3>
                </div>
                <div
                  style={
                    Object {
                      "boxSizing": "border-box",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "left": "auto",
                      "marginLeft": "0%",
                      "maxWidth": "100%",
                      "minHeight": 1,
                      "paddingLeft": 15,
                      "paddingRight": 15,
                      "position": "relative",
                      "right": "auto",
                      "width": "100%",
                    }
                  }
                >
                  <p
                    style={
                      Object {
                        "marginTop": 10,
                      }
                    }
                  >
                    For example, your card number, PIN and PUK of your mobile phone, the serial number of any of your devices or any information you need to keep in a safe place.
                  </p>
                </div>
              </div>
              <div
                style={
                  Object {
                    "alignItems": "normal",
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "flexWrap": "wrap",
                    "justifyContent": "flex-start",
                    "marginLeft": -15,
                    "marginRight": -15,
                    "marginTop": 25,
                  }
                }
              >
                <div
                  style={
                    Object {
                      "boxSizing": "border-box",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "left": "auto",
                      "marginLeft": "0%",
                      "maxWidth": "100%",
                      "minHeight": 1,
                      "paddingLeft": 15,
                      "paddingRight": 15,
                      "position": "relative",
                      "right": "auto",
                      "width": "100%",
                    }
                  }
                >
                  <div
                    className="Checkbox-container"
                  >
                    <input
                      checked={false}
                      className="Checkbox"
                      id="Checkbox"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <label
                      className="Checkbox-label"
                      htmlFor="Checkbox"
                    >
                      I'm of legal age and I accept that my data will be processed according to the Data Protection Policy.
                    </label>
                  </div>
                </div>
              </div>
              <span
                style={
                  Object {
                    "clear": "both",
                    "display": "table",
                  }
                }
              />
            </div>
          </div>
          <div
            className="ButtonContainer"
          >
            <div
              style={
                Object {
                  "boxSizing": "border-box",
                  "height": "100%",
                  "marginLeft": "auto",
                  "marginRight": "auto",
                  "paddingLeft": 15,
                  "paddingRight": 15,
                  "position": "relative",
                }
              }
            >
              <div
                style={
                  Object {
                    "alignItems": "center",
                    "display": "flex",
                    "flexGrow": 0,
                    "flexShrink": 0,
                    "flexWrap": "wrap",
                    "height": "100%",
                    "justifyContent": "space-between",
                    "marginLeft": -15,
                    "marginRight": -15,
                  }
                }
              >
                <div
                  className="ButtonContainer-submit"
                  style={
                    Object {
                      "boxSizing": "border-box",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 0,
                      "left": "auto",
                      "marginLeft": "0%",
                      "maxWidth": "100%",
                      "minHeight": 1,
                      "paddingLeft": 15,
                      "paddingRight": 15,
                      "position": "relative",
                      "right": "auto",
                      "width": "100%",
                    }
                  }
                >
                  <button
                    className="Button-filled color-secondary"
                    disabled={true}
                    onClick={[Function]}
                  >
                    Next
                    <div
                      className="Button-endAddornment"
                    >
                      <svg
                        fill="#fff"
                        height="24"
                        width="24"
                      >
                        arrow_right.svg
                      </svg>
                    </div>
                  </button>
                </div>
              </div>
              <span
                style={
                  Object {
                    "clear": "both",
                    "display": "table",
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</main>
`;
